Add body_format for handling of JSON and YAML body

sbocahu opened the pull request:
body_format is a new optional argument that enables handling of JSON or
YAML serialization format for the body argument.
When set to either 'json' or 'yaml', the body argument can be a dict or list.
The body will be encoded, and the Content-Type HTTP header will be set,
accordingly to the body_format argument.

Example:

```
- name: Facette - Create memory graph
  uri:
    method: POST
    url: http://facette/api/v1/library/graphs
    status_code: 201
    body_format: json
    body:
      name: "{{ ansible_fqdn }} - Memory usage"
      attributes:
        Source": "{{ ansible_fqdn }}"
        link: "1947a490-8ac6-4bf2-47c1-ff74272f8b32"
```

Commit:
Add body_format for handling of JSON and YAML body

body_format is a new optional argument that enables handling of JSON or
YAML serialization format for the body argument.
When set to either 'json' or 'yaml', the body argument can be a dict or list.
The body will be encoded, and the Content-Type HTTP header will be set,
accordingly to the body_format argument.

Example:

- name: Facette - Create memory graph
  uri:
    method: POST
    url: http://facette/api/v1/library/graphs
    status_code: 201
    body_format: json
    body:
      name: "{{ ansible_fqdn }} - Memory usage"
      attributes:
        Source": "{{ ansible_fqdn }}"
        link: "1947a490-8ac6-4bf2-47c1-ff74272f8b32"

Doooooo0o commented:
+1

dkerwin commented:
:+1:

sivel commented on network/basics/uri.py:
```diff
@@ -367,14 +374,24 @@ def main():
+        body = json.dumps(body)
+        dict_headers['Content-Type'] = 'application/json'
+    elif body_format == 'yaml':
+        body = yaml.dump(body)
```
The YAML module is never imported anywhere, additionally the YAML python module is not a required module for a ansible managed machine.

Does this provide much benefit?  I have never seen an API accept `application/yaml` anyway.

sbocahu replied:
That's right.

I only added this option "for others" and "in case of" but I never seen such an API neither :-)
I removed YAML from my PR.

Thanks for commenting.

bcoca commented:
even if an API takes YAML, it will take the json as it is a subset.

You can keep the yaml option, but state clearly the requirement and make sure you import it.

smiller171 commented:
:+1:

Commit:
Removes YAML body_format option

It would require adding a dependency to the YAML module while it might
not provide much benefit (has anybody seen a HTTP API taking YAML as
input ?)

F21 commented:
When will this be released?

sbocahu commented:
@F21 that's already merged.

ghost commented:
@zecrazytux Can you identify the release? I'm running Ansible 1.9.0.1 and I'm getting this error when I try to post a JSON body to a web service.

smiller171 commented:
@zecrazytux This was merged 5 days after the 1.9.0.1 release. Unfortunately I don't see the commit in the current stable-1.9 branch either so you won't find it in 1.9.1

bcoca commented:
1.9.1 was a bugfix release, new features don't normally get added to minor releases.

smiller171 commented:
@bcoca This is a bugfix. Currently passing Json to an API endpoint using a template fails with a very unhelpful stacktrace error. This fixes that.

bcoca commented:
its a new parameter that enables new functionality, the stacktrace is still a bug.

smiller171 commented:
I would argue that the inability to use variables in json passed with the uri module is a bug in it's own right.

ghost commented:
@smiller171 @bcoca Agree that this is a bug, but it definitely wasn't included in 1.9.1 :(

smiller171 commented:
Well using curl through the command or shell modules will continue to work while we wait. @ruebrenda

jedelman8 commented:
@abadger @bcoca So, this worked for a basic JSON datastructure, but now getting another issue when it's a nested data structure, i.e. dictionary is a value.

With this as the body:
```
body: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": "{{ site }}" } '
```

It raises this error:
```
failed: [localhost] => (item=one) => {"failed": true, "item": "one", "parsed": false}
Traceback (most recent call last):
  File "/home/cisco/.ansible/tmp/ansible-tmp-1443751225.39-78407924015857/uri", line 2067, in <module>
    main()
  File "/home/cisco/.ansible/tmp/ansible-tmp-1443751225.39-78407924015857/uri", line 419, in main
    resp, content, dest = uri(module, url, dest, user, password, body, method, dict_headers, redirects, socket_timeout, validate_certs)
  File "/home/cisco/.ansible/tmp/ansible-tmp-1443751225.39-78407924015857/uri", line 312, in uri
    resp, content = h.request(url, method=method, body=body, headers=headers)     
  File "/usr/lib/python2.7/dist-packages/httplib2/__init__.py", line 1569, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "/usr/lib/python2.7/dist-packages/httplib2/__init__.py", line 1316, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "/usr/lib/python2.7/dist-packages/httplib2/__init__.py", line 1252, in _conn_request
    conn.request(method, request_uri, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 979, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1013, in _send_request
    self.endheaders(body)
  File "/usr/lib/python2.7/httplib.py", line 975, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 839, in _send_output
    self.send(message_body)
  File "/usr/lib/python2.7/httplib.py", line 811, in send
    self.sock.sendall(data)
  File "/usr/lib/python2.7/ssl.py", line 329, in sendall
    v = self.send(data[count:])
TypeError: unhashable type


```

But then adding `to_json` like so:

```
body: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": "{{ site }}" } | to_json '
```

Produces this error:
```
failed: [localhost] => (item=one) => {"allow": "GET, POST, HEAD, OPTIONS", "connection": "close", "content": "{\"detail\": \"JSON parse error - Extra data: line 1 column 124 - line 1 column 134 (char 123 - 133)\"}", "content_type": "application/json", "date": "Fri, 02 Oct 2015 02:01:36 GMT", "failed": true, "item": "one", "json": {"detail": "JSON parse error - Extra data: line 1 column 124 - line 1 column 134 (char 123 - 133)"}, "redirected": false, "server": "Apache/2.4.7 (Ubuntu)", "status": 400, "transfer_encoding": "chunked", "vary": "Accept,Cookie", "x_api_time": "0.001s"}
msg: Status code was not [201]
```

Have tried everything possible, but still getting JSON parse error as soon as a variable is being inserted into a nested json/dict, etc.

Thoughts?

jedelman8 commented:
@abadger @bcoca still getting an error when a variable is being inserted within a nested dictionary like so with the site key in extra_vars :

```
     - uri:
         url: https://192.168.56.102/api/v1/jobs/
         method: POST
         body_format: json
         body: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": { "site": "{{ site }}" } } | to_json'
         status_code: 201
         HEADER_Content-Type: "application/json"
         validate_certs: no
         user: admin
         password: ansible
```

This produces the following JSON parse error:

```
failed: [localhost] => (item=one) => {"allow": "GET, POST, HEAD, OPTIONS", "connection": "close", "content": "{\"detail\": \"JSON parse error - Extra data: line 1 column 136 - line 1 column 145 (char 135 - 144)\"}", "content_type": "application/json", "date": "Fri, 02 Oct 2015 02:07:46 GMT", "failed": true, "item": "one", "json": {"detail": "JSON parse error - Extra data: line 1 column 136 - line 1 column 145 (char 135 - 144)"}, "redirected": false, "server": "Apache/2.4.7 (Ubuntu)", "status": 400, "transfer_encoding": "chunked", "vary": "Accept,Cookie", "x_api_time": "0.001s"}
msg: Status code was not [201]
```

It works fine if extra_vars is removed OR if site is not referencing a variable as can be seen here:

```
     - uri:
         url: https://192.168.56.102/api/v1/jobs/
         method: POST
         body_format: json
         body: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": { "site": "testing" } }'
         status_code: 201
         HEADER_Content-Type: "application/json"
         validate_certs: no
         user: admin
         password: ansible
```

and good output:

```
failed: [localhost] => (item=one) => {"allow": "GET, POST, HEAD, OPTIONS", "connection": "close", "content": "{\"job_template\": \"Invalid job template\"}", "content_type": "application/json", "date": "Fri, 02 Oct 2015 02:10:05 GMT", "failed": true, "item": "one", "json": {"job_template": "Invalid job template"}, "redirected": false, "server": "Apache/2.4.7 (Ubuntu)", "status": 400, "transfer_encoding": "chunked", "vary": "Accept,Cookie", "x_api_time": "0.111s"}
msg: Status code was not [201]

```

Because it is an invalid job template...and the good news is this is testing the API on Tower :)

Thoughts?

bcoca commented:
have you tried just creating the structure in yaml in a var and then using:

```
body: "{{varname|to_json}}"
```

jedelman8 commented:
Just tested these vars with `body: "{{ v2 | to_json }}"`:

```
   vars:
     username: admin
     password: ansible
     site: abcd
     v1: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": { "site": "{{ site }}" } }'
     v2: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": { "site": "abcd" } }'
     v3: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41 }'
```

using v1 yields (and this is what I really want): 

```
TypeError: unhashable type

```

using v2 and v3 yield:

```
failed: [localhost] => (item=one) => {"allow": "GET, POST, HEAD, OPTIONS", "connection": "close", "content": "{\"non_field_errors\": [\"Invalid data\"]}", "content_type": "application/json", "date": "Fri, 02 Oct 2015 03:30:32 GMT", "failed": true, "item": "one", "json": {"non_field_errors": ["Invalid data"]}, "redirected": false, "server": "Apache/2.4.7 (Ubuntu)", "status": 400, "transfer_encoding": "chunked", "vary": "Accept,Cookie", "x_api_time": "0.055s"}
msg: Status code was not [201]
```

bcoca commented:
not what i meant:

```
vars:
  v1:
    name: "Add a Customer"
    description: "add a customer"
    job_type: "run"
    job_template: 41
    extra_vars: 
      site: "{{ site }}" 
  v2:
    name: "Add a Customer"
    description: "add a customer"
    job_type: "run"
    job_template: 41
    extra_vars: 
      site: "abcd" 
```

jedelman8 commented:
No, they both provide an unhashable type error.

jedelman8 commented:
Could the problem be Ansible Tower?

This seems to work on another REST server:

```
         url: http://apic/api/aaaLogin.json
         method: POST
         body_type: json
         body: '{ "aaaUser": { "attributes": { "name": "{{ item }}", "pwd": "C1sco12345" } } } | to_json'
```

But this:

```
         url: https://192.168.56.102/api/v1/jobs/
         method: POST
         body: '{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", "job_template": 41, "extra_vars": { "site": "{{ site }}" } } | to_json '
         body_format: json
         # body: "{{ v1 | to_json }}"
         status_code: 201
         HEADER_Content-Type: "application/json"
         validate_certs: no
         user: admin
         password: ansible
```

produces the error on having "extra data" as seen below...and this error IS IN FACT coming from Tower 2.1.1.

This also validates the rendering of vars is good...and this uses to_json at the end of the complete string.

```
<localhost> REMOTE_MODULE uri method=POST password=VALUE_HIDDEN body_format=json
 user=admin url=https://192.168.56.102/api/v1/jobs/ HEADER_Content-Type=application/json 
body='{ "name": "Add a Customer", "description": "add a customer", "job_type": "run", 
"job_template": 41, "extra_vars": { "site": "abcd" } } | to_json '

failed: [localhost] => (item=one) => {"allow": "GET, POST, HEAD, OPTIONS", "connection": "close",
 "content": "{\"detail\": \"JSON parse error - Extra data: line 1 column 136 - line 1 column 146 (char 
135 - 145)\"}", "content_type": "application/json", "date": "Fri, 02 Oct 2015 11:11:49 GMT", "failed":
 true, "item": "one", "json": {"detail": "JSON parse error - Extra data: line 1 column 136 - line 1 
column 146 (char 135 - 145)"}, "redirected": false, "server": "Apache/2.4.7 (Ubuntu)", "status": 400, 
"transfer_encoding": "chunked", "vary": "Accept,Cookie", "x_api_time": "0.001s"}
```

Thoughts?

willthames commented:
ok, the problem is nothing to do with Ansible Tower, @bcoca's suggestion of using `to_json` does not help.

I'm not even convinced this is a bug in the uri module, even though the uri module is one of the most likely places you will encounter this.

This is because after the `to_json` filter plugin is called, the result is passed back to template. Because the string starts with '{', the string is then passed to safe_eval, which converts it back from a string to a dict. Completely eliminating the point of `to_json`. Is `to_json` confirmed as actually having an effect?

```
-> complex_args = template.template(self.basedir, complex_args, inject, fail_on_undefined=self.error_on_undefined_vars)
  /home/wthames/src/ansible/lib/ansible/utils/template.py(140)template()
-> d[k] = template(basedir, v, templatevars, lookup_fatal, depth, expand_lists, convert_bare, fail_on_undefined, filter_fatal)
  /home/wthames/src/ansible/lib/ansible/utils/template.py(124)template()
-> varname = template_from_string(basedir, varname, templatevars, fail_on_undefined)
  /home/wthames/src/ansible/lib/ansible/utils/template.py(382)template_from_string()
-> res = jinja2.utils.concat(rf)
  <template>(9)root()
> /home/wthames/src/ansible/lib/ansible/runner/filter_plugins/core.py(58)to_json()
-> return result
```

The problematic code is 
https://github.com/ansible/ansible/blob/stable-1.9/lib/ansible/utils/template.py#L121-L133

After `template_from_string` is called, returning the correct json serialization of the variable, the next thing that happens is that `utils.safe_eval` is called, turning it back to a dict. 

I've just hit this with 1.9.4

This is enough of a test case

```
- hosts: localhost
  connection: local
  gather_facts: False
  vars:
    test_dict:
      hello: world

  tasks:
  - debug: var=test_dict

  - set_fact: 
    args:
      test_dict_to_json: "{{test_dict|to_json}}"

  - debug: var=test_dict_to_json
```

We expect `test_dict_to_json` to look a bit like '{"hello": "world"}'

willthames commented:
Note that this bug where `to_json` is pointless is still in Ansible 2.0

bcoca commented:
known, this is the 'convert to known data type if possible' in template function, it should be skipped when template ends in |to_json, |to_nice_json or , |pretty, for now the work around is adding an starting space to the templated string, which will bypass the 'typization'
